### PR TITLE
Schedule Modal Refactor

### PR DESCRIPTION
**Pull request by holderb5**
Resolves #

## Proposed Changes

-

## Screenshots
Screen Shot 2022-05-12 at 9 04 27 PM

Screen Shot 2022-05-12 at 9 05 22 PM
Screen Shot 2022-05-12 at 9 05 03 PM

Screen Shot 2022-05-13 at 2 07 30 PM

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: packages/core/addon/models/delivery-rule.ts [hunk 1]
```diff
@@ -64,6 +64,9 @@ export default class DeliveryRuleModel extends Model.extend(Validations) {
   @fragment('fragments/delivery-format', { polymorphic: true })
   format!: DeliveryFormatFragment;
 
+  @attr({ defaultValue: 'Email delivered pdf every week' })
```
[the remark this answers is not in the document]
Comment (holderb5) [reply packages/core/addon/models/delivery-rule.ts:67]: I was trying to make it dynamic based on the other values but because they might not be defined I left it at the default values. Do you have a suggestion on how to do a better default?

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 2]
```diff
@@ -72,6 +72,8 @@ export default class ScheduleActionComponent extends Component<Args> {
 
   @tracked showModal = false;
 
+  @tracked errorWhileFetchingRules!: boolean;
```
[the remark this answers is not in the document]
Comment (holderb5) [reply packages/reports/addon/components/common-actions/schedule.ts:75]: Yup is used in the template

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 3]
```diff
@@ -142,80 +144,105 @@ export default class ScheduleActionComponent extends Component<Args> {
...
-        owner: this.user.getUser(),
-      });
-
-      this.attemptedSave = false;
-      let savePromise = new RSVP.Promise((resolve, reject) => {
-        assert('The localDeliveryRule is defined', this.localDeliveryRule);
-        this.args.onSave(this.localDeliveryRule, { resolve, reject });
-      });
+    this.deliveryRules?.forEach(async (rule) => {
+      if (rule.validations.isValid) {
```
[the remark this answers is not in the document]
Comment (holderb5) [reply packages/reports/addon/components/common-actions/schedule.ts:176]: Ahh I'll have to change it to check if all rules are valid not just the active one. Good catch

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 4]
```diff
@@ -142,80 +144,105 @@ export default class ScheduleActionComponent extends Component<Args> {
...
       //Add Page notification
       this.naviNotifications.add({
-        title: `Delivery schedule removed`,
+        title: `Delivery schedule '${name}' removed`,
         style: 'success',
         timeout: 'short',
       });
+
+      if (!this.deliveryRules || this.deliveryRules.length <= 0) {
+        this.closeModal();
```
[the remark this answers is not in the document]
Comment (holderb5) [reply packages/reports/addon/components/common-actions/schedule.ts:242]: Yes do we not want to do that? I can just remove this

File: packages/reports/addon/components/common-actions/schedule.hbs [hunk 5]
```diff
@@ -7,207 +7,279 @@
...
+              <DenaliIcon @icon="add-circle" />
+              New Schedule
             </DenaliButton>
-          </CommonActions::Delete>
-        {{/unless}}
-      {{/if}}
+          </div>
+          <DenaliSidebar class="has-bg-white w-full" as |Sidebar|>
+            {{#each this.deliveryRules as |deliveryRule|}}
+              <Sidebar.Tab {{on "click" (fn (mut this.localDeliveryRule) deliveryRule)}}>
```
[the remark this answers is not in the document]
Comment (holderb5) [reply packages/reports/addon/components/common-actions/schedule.hbs:34]: What do you mean by this? How do you want that to work?

File: packages/reports/addon/components/common-actions/schedule.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkusa) [reply packages/reports/addon/components/common-actions/schedule.ts:75]: Ah, the diff was collapsed 👍

File: packages/reports/addon/components/common-actions/schedule.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkusa) [reply packages/reports/addon/components/common-actions/schedule.ts:238]: yeah, I think that might be unexpected to the user

File: packages/reports/addon/components/common-actions/schedule.hbs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkusa) [reply packages/reports/addon/components/common-actions/schedule.hbs:34]: When a rule is selected, the text changes color and has a boarder around it: 
Screen Shot 2022-03-15 at 4 23 51 PM

File: packages/reports/addon/components/common-actions/schedule.hbs [hunk 6]
```diff
@@ -7,207 +7,282 @@
...
+                </DenaliButton>
+              {{else}}
+                <DenaliButton
+                  class="schedule__modal-save-btn m-r-5"
+                  disabled={{this.disableSave}}
+                  isLoading={{this.isSaving}}
+                  {{on "click" (queue (toggle "isSaving" this) this.doSave)}}
+                >
+                  {{#if this.localDeliveryRule.isNew}}
+                    Save
```
Comment (100starnight) [inline packages/reports/addon/components/common-actions/schedule.hbs:246]: If you wind up doing the "clicking on save means that the user is saving all of the schedules" thing, it might make more sense to move the save button to the left panel under the list of schedules

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 7]
```diff
@@ -290,7 +290,9 @@ export default class ScheduleActionComponent extends Component<Args> {
     assert('The localDeliveryRule is defined', this.localDeliveryRule);
     assert('The format is defined', this.localDeliveryRule.format);
     this.localDeliveryRule.format.type = type;
-    if (this.localDeliveryRule.name.toString().match(/\w+ \bdelivered\b +\w+ \bevery\b +\w/)) {
+    if (
+      this.localDeliveryRule.name.toString().match(/\w+ \bdelivered\b +\w+ \bevery\b (day|week|month|quarter|year)$/)
```
Comment (100starnight) [inline packages/reports/addon/components/common-actions/schedule.ts:294]: This is still acting funky - for example, if you add a number to the end of the default name, for example, the next time you open the modal it only have the default name and not the number the user added

Comment (holderb5) [reply packages/reports/addon/components/common-actions/schedule.ts:301]: That's because I had something in onOpen that was replacing it, its working fine now

File: packages/reports/addon/components/common-actions/schedule.hbs [hunk 8]
```diff
@@ -7,207 +7,282 @@
...
+                  {{#if (includes this.localDeliveryRule.format.type this.overwriteableFormats)}}
+                    <DenaliInputGroup @label="Overwrite spreadsheet" class="schedule__modal-overwrite">
+                      <DenaliSwitch
+                        class="schedule__modal-overwrite-toggle"
+                        @isEnabled={{this.localDeliveryRule.format.options.overwriteFile}}
+                        @onChange={{this.toggleOverwriteFile}}/>
+                    </DenaliInputGroup>
+                    <p class="help-text m-b-30 is-grey-800">
+                      {{#if this.localDeliveryRule.format.options.overwriteFile}}
+                        Data will be replaced in the same sheet every {{this.localDeliveryRule.frequency}}
```
[the remark this answers is not in the document]
Comment (cythrawll) [reply packages/reports/addon/components/common-actions/schedule.hbs:195]: it shouldn't be an issue anymore. however, messing with the sheet will have their stuff being overwritten. But would be a bit too wordy to put here. Also it's pretty specific to our export library and requirements so probably shouldn't be mentioned here in framework.

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 9]
```diff
@@ -134,88 +144,105 @@ export default class ScheduleActionComponent extends Component<Args> {
       return true;
     }
 
-    return !(this.localDeliveryRule?.hasDirtyAttributes && this.isRuleValid);
+    return !(this.localDeliveryRule?.hasDirtyAttributes && this.areRulesValid);
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.ts:147]: Did we move do individual saves? If so, do we need to check if all are valid to save one rule?

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 10]
```diff
@@ -122,9 +123,18 @@ export default class ScheduleActionComponent extends Component<Args> {
...
-  @reads('localDeliveryRule.validations.isValid') isRuleValid!: boolean;
+  get areRulesValid() {
+    let valid = true;
+    this.deliveryRules.forEach(async (rule) => {
+      if (!rule.validations.isValid) {
+        valid = false;
+      }
+    });
+
+    return valid;
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.ts:136]:
```suggestion
    return this.this.deliveryRules.every((r) => r.validations.isValid);
```

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 11]
```diff
@@ -134,88 +144,105 @@ export default class ScheduleActionComponent extends Component<Args> {
...
   _createNewDeliveryRule() {
-    return this.store.createRecord('delivery-rule', {
+    let newRule = this.store.createRecord('delivery-rule', {
       format: { type: this.formats.firstObject },
+      deliveredItem: this.args.model,
+      owner: this.user.getUser(),
     });
+
+    assert('Delivery Rules are defined', this.deliveryRules);
+    this.deliveryRules = [...this.deliveryRules, newRule];
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.ts:176]: Can we move this elsewhere? We this seems like a side effect to a method called `_createNewDeliveryRule`

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 12]
```diff
@@ -134,88 +144,105 @@ export default class ScheduleActionComponent extends Component<Args> {
...
    * @property {Boolean} attemptedSave
    */
   @tracked attemptedSave = false;
 
+  /**
+   * @action addNewRule
+   */
+  @action
+  addNewRule() {
+    this.localDeliveryRule = this._createNewDeliveryRule();
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.ts:160]: What does `localDeliveryRule` do now that we can have multiple local rules? Should be this `currentDisplyedRule`?

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 13]
```diff
@@ -134,88 +144,105 @@ export default class ScheduleActionComponent extends Component<Args> {
...
-          title: `${capitalize(this.localDeliveryRule.deliveryType)} delivery schedule successfully saved!`,
-          style: 'success',
-          timeout: 'short',
+    this.deliveryRules?.forEach(async (rule) => {
+      if (rule.validations.isValid) {
+        // Only add relationships to the new delivery rule if the fields are valid
+        this.attemptedSave = false;
+        let savePromise = new RSVP.Promise((resolve, reject) => {
+          assert('The Rule is defined', rule);
+          this.args.onSave(rule, { resolve, reject });
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.ts:191]: Given the changes to the flow, do we still need to do a callback on save?

Comment (jkusa) [reply packages/reports/addon/components/common-actions/schedule.ts:191]: Could something like this be done 
```ts
try {
  await this.args.saveRule(rule);
} catch (e) {
  //handle error
}
```

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 14]
```diff
@@ -134,88 +144,105 @@ export default class ScheduleActionComponent extends Component<Args> {
...
-  async doDelete() {
-    const deliveryRule = this.localDeliveryRule;
+  async doDelete(deliveryRule: DeliveryRuleModel) {
     assert('The localDeliveryRule is defined', deliveryRule);
+    const name = deliveryRule.name;
     const deletePromise = new RSVP.Promise((resolve, reject) => {
       this.args.onDelete(deliveryRule, { resolve, reject });
     });
 
+    this.deliveryRules = this.deliveryRules?.filter((rule) => rule !== deliveryRule);
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.ts:230]: What if the delete fails? Is this being done too soon? Do we need this? Is  `deliveryRulesForUser` not getting recomputed after the delete?

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 15]
```diff
@@ -224,14 +251,29 @@ export default class ScheduleActionComponent extends Component<Args> {
...
   async onOpen() {
-    //Kick off a fetch for existing delivery rules
-    this.deliveryRule = this.args.model.deliveryRuleForUser;
-
     try {
-      const rule = await this.deliveryRule;
-      this.localDeliveryRule = rule ? rule : this.localDeliveryRule || this._createNewDeliveryRule();
+      this.deliveryRules = await this.args.model.deliveryRulesForUser;
+      if (this.deliveryRules) {
+        this.localDeliveryRule = this.deliveryRules[0] ? this.deliveryRules[0] : this.localDeliveryRule || undefined;
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.ts:257]: Would `localDeliveryRule` be defined at this point?

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 16]
```diff
@@ -224,14 +251,29 @@ export default class ScheduleActionComponent extends Component<Args> {
...
+        if (this.localDeliveryRule) {
+          if (
+            this.localDeliveryRule.format.type !== 'pdf' &&
+            this.localDeliveryRule.delivery !== 'none' &&
+            this.localDeliveryRule.name
+              .toString()
+              .match(/\w+ \bdelivered\b +\w+ \bevery\b (day|week|month|quarter|year)$/)
+          ) {
+            this.localDeliveryRule.name = `Email delivered ${this.localDeliveryRule.format.type} every week`;
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.ts:266]: Will this override an existing name?

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 17]
```diff
@@ -224,14 +251,29 @@ export default class ScheduleActionComponent extends Component<Args> {
...
+          }
+          if (this.formats.length === 1) {
+            this.updateFormat(this.formats[0]);
+          }
+        }
+      }
     } catch (e) {
+      this.errorWhileFetchingRules = true;
       this.localDeliveryRule = this._createNewDeliveryRule();
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.ts:275]: Do we need to create a new rule when fetching fails?

File: packages/reports/addon/components/common-actions/schedule.ts (hunk 16 above)
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.ts:266]: I think it might be easier to not store the generated string. Can we generate the string on the client in the case a user defined name is empty?

File: packages/reports/addon/components/common-actions/schedule.ts (hunk 9 above)
Comment (holderb5) [reply packages/reports/addon/components/common-actions/schedule.ts:147]: We didnt move to individual saves

File: packages/reports/addon/components/common-actions/schedule.ts (hunk 12 above)
Comment (holderb5) [reply packages/reports/addon/components/common-actions/schedule.ts:160]: Yeah that would be a better name, I'll swap it.

File: packages/reports/addon/components/common-actions/schedule.ts (hunk 16 above)
Comment (holderb5) [reply packages/reports/addon/components/common-actions/schedule.ts:266]: So whenever the customer sets the delivery rule name to an empty string we just pop in the default? And by stored on client what do you mean?

Comment (jkusa) [reply packages/reports/addon/components/common-actions/schedule.ts:266]: Yes, when the name `isBlank` render the generated name in the component.

File: packages/reports/addon/components/common-actions/schedule.ts (hunk 11 above)
Comment (jkusa) [reply packages/reports/addon/components/common-actions/schedule.ts:176]: line 176

File: packages/reports/addon/components/common-actions/schedule.hbs [hunk 18]
```diff
@@ -2,212 +2,287 @@
...
+                      class="schedule__modal-disabled-status"
+                      @isEnabled={{not this.currentDisplayedRule.isDisabled}}
+                      @onChange={{toggle "isDisabled" this.currentDisplayedRule}}/>
+                    <p>
+                      {{#if this.currentDisplayedRule.isDisabled}}
+                        Paused
+                      {{else}}
+                        Active
+                      {{/if}}
+                    </p>
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.hbs:212]:
```suggestion
                      @onChange={{toggle "isDisabled" this.currentDisplayedRule}}
                      @onLabel="Active"
                      @offLabel="Paused"
                    />
```

File: packages/reports/addon/components/common-actions/schedule.hbs [hunk 19]
```diff
@@ -2,212 +2,287 @@
...
+                  {{on "click" (queue this.closeModal this.revertAll)}}
+                >
+                  {{#if this.currentDisplayedRule.hasDirtyAttributes}}
+                    Cancel
+                  {{else}}
+                    Close
+                  {{/if}}
+                </DenaliButton>
+              {{/if}}
+              {{#unless this.currentDisplayedRule.isNew}}
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.hbs:265]: Can we clean this up? This isn't in the design spec.

File: packages/reports/addon/components/common-actions/schedule.hbs [hunk 20]
```diff
@@ -2,212 +2,287 @@
...
+                  isLoading={{this.isSaving}}
+                  {{on "click" (queue (toggle "isSaving" this) this.doSave)}}
+                >
+                  {{#if this.currentDisplayedRule.isNew}}
+                    Save
+                  {{else}}
+                    Save Changes
+                  {{/if}}
+                </DenaliButton>
+                <DenaliButton
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.hbs:251]: Can we clean this up? This isn't in the design spec.

File: packages/reports/addon/components/common-actions/schedule.hbs [hunk 21]
```diff
@@ -2,212 +2,287 @@
...
   class="schedule__modal"
   ...attributes
   as |Modal|
 >
-  <Modal.Content>
-    <h3 class="schedule__modal-header m-b-10">Schedule {{capitalize this.modelType}}</h3>
-    {{#if (or (is-pending this.isValidForSchedule) (is-pending this.deliveryRule))}}
-      <NaviLoader />
+  <Modal.Header>
+    <h4 class="schedule__modal-header">Schedule {{capitalize this.modelType}}</h4>
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.hbs:11]:
```suggestion
    <h3 class="schedule__modal-header">Schedule {{capitalize this.modelType}}</h3>
```
We've been mainly using h3 here, any reason to make this h4?

File: packages/reports/addon/components/common-actions/schedule.hbs [hunk 22]
```diff
@@ -2,212 +2,287 @@
...
+              >
+                <Alert.Context>
+                  Unable to schedule invalid {{this.modelType}}. Please fix errors before proceeding.
+                </Alert.Context>
+              </DenaliAlert>
+            {{else}}
+                {{#if (eq this.currentDisplayedRule undefined)}}
+                  <div class="schedule__modal-no-schedule">
+                    <DenaliIcon class="flex justify-content-center m-t-200" @size="large" @icon="calendar-event" />
+                    <h3 class="flex justify-content-center">We Deliver For You!</h3>
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.hbs:81]:
```suggestion
                    <h3 class="flex justify-content-center">Let’s jump right in!</h3>
```
The design team updated the copy

**Comment by jkusa** [comment]
Looks like we need to allow the left rail to scroll
Screen Shot 2022-04-25 at 1 29 40 PM

**Comment by jkusa** [comment]
Looks like labels are out of sync for new schedules and for some reason the format selector is disabled:

Screen Shot 2022-04-25 at 1 37 20 PM

File: packages/reports/addon/components/common-actions/schedule.hbs [hunk 23]
```diff
@@ -2,212 +2,266 @@
...
+                      as |onDelete|
+                    >
+                      <DenaliButton
+                        disabled={{this.isSaving}}
+                        @style="text"
+                        @size="medium"
+                        class="schedule__modal-delete-btn"
+                        {{on "click" onDelete}}
+                      >
+                        Delete
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.hbs:256]: Was this in the spec? The delete behavior is different then the trash can on the left.

File: packages/reports/addon/components/common-actions/schedule.ts [hunk 24]
```diff
@@ -134,88 +144,102 @@ export default class ScheduleActionComponent extends Component<Args> {
...
   /**
-   * @property {Boolean} attemptedSave
+   * @action addNewRule
    */
-  @tracked attemptedSave = false;
+  @action
+  addNewRule() {
+    this.currentDisplayedRule = this._createNewDeliveryRule();
+    if (this.currentDisplayedRule.name === '') {
+      this.currentDisplayedRule.name = this.defaultName;
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.ts:157]: Instead of saving the default rule, can we just display the `defaultName` in the case the model name is blank/empty?

File: packages/reports/addon/components/common-actions/schedule.hbs [hunk 25]
```diff
@@ -92,122 +92,124 @@
                     </DenaliButton>
                   </div>
                 {{else}}
-                  <DenaliInputGroup @isStacked={{true}} @label="Schedule Name" class="m-b-30">
-                    <DenaliInput 
-                      @wrapperClass="w-full" 
-                      type="text" 
-                      placeholder="Schedule Name (Optional)"
-                      class="schedule__modal-name"
-                      value={{unless this.currentDisplayedRule.name this.currentDisplayedRule.name this.defaultName}}
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.hbs]: Can we do this approach? A computed property could be used as well. Also, do we have tests around the bug we had before?

Comment (jkusa) [reply packages/reports/addon/components/common-actions/schedule.hbs]: Yeah, you could have a computed property with a getter and setting backed by the persisted `name` attribute

File: packages/core/addon/models/delivery-rule.ts [hunk 26]
```diff
@@ -45,6 +46,11 @@ const Validations = buildValidations({
 });
 
 export default class DeliveryRuleModel extends Model.extend(Validations) {
+  get defaultName() {
+    return `${capitalize(this.delivery)} delivered ${this.format.type} every ${this.frequency}`;
+  }
+
+  nameChanged = false;
```
Comment (jkusa) [inline packages/core/addon/models/delivery-rule.ts:53]: Is this used?

File: packages/reports/app/styles/navi-reports/components/common-actions/schedule-modal.scss [hunk 27]
```diff
@@ -58,19 +65,31 @@
     &-delivery-list {
       border-right: 1px solid #e8e8e8;
 
+      li {
+        height: fit-content !important;
```
Comment (jkusa) [inline packages/reports/app/styles/navi-reports/components/common-actions/schedule-modal.scss:69]: why the `!important`?

File: packages/reports/index.js [hunk 28]
```diff
@@ -1,4 +1,8 @@
 'use strict';
+// eslint-disable-next-line node/no-extraneous-require
+let Funnel = require('broccoli-funnel');
+// eslint-disable-next-line node/no-extraneous-require
```
Comment (jkusa) [inline packages/reports/index.js:4]: Why do we have these disable lints? Did we add `broccoli-merge-trees` & `broccoli-funnel` as dependencies in the package.json?

**Review by jkusa — COMMENTED** [review]
Looks like we need to update the name logic here:
Screen Shot 2022-05-12 at 12 24 09 PM

Look like we have some responsive issues: 
Screen Shot 2022-05-12 at 12 37 43 PM

File: packages/reports/app/styles/navi-reports/components/common-actions/schedule-modal.scss [hunk 29]
```diff
@@ -40,5 +48,65 @@
...
+      position: relative;
+
+      &-inputs {
+        height: 455px;
+        overflow-y: auto;
+      }
+    }
+
+    &-delivery-list {
+      border-right: 1px solid #e8e8e8;
```
Comment (kevinhinterlong) [inline packages/reports/app/styles/navi-reports/components/common-actions/schedule-modal.scss:66]:
```suggestion
      border-right: 1px solid map-get($denali-grey-colors, '400');
```

File: packages/reports/app/styles/navi-reports/components/common-actions/schedule-modal.scss [hunk 30]
```diff
@@ -40,5 +48,65 @@
...
+      left: 50%;
+      margin-top: 50px;
+      position: absolute;
+      top: 50%;
+      transform: translate(-50%, -50%);
+      width: 300px;
+    }
+
+    &-buttons {
+      border-top: 1px solid #e8e8e8;
```
Comment (kevinhinterlong) [inline packages/reports/app/styles/navi-reports/components/common-actions/schedule-modal.scss:98]:
```suggestion
      border-top: 1px solid map-get($denali-grey-colors, '400');
```

File: packages/reports/app/styles/navi-reports/components/common-actions/schedule-modal.scss [hunk 31]
```diff
@@ -40,5 +48,65 @@
...
+    }
+
+    &-new-delivery {
+      border-bottom: 1px solid #e8e8e8;
```
Comment (kevinhinterlong) [inline packages/reports/app/styles/navi-reports/components/common-actions/schedule-modal.scss:102]:
```suggestion
      border-bottom: 1px solid map-get($denali-grey-colors, '400');
```

File: packages/webservice/models/src/main/kotlin/com/yahoo/navi/ws/models/beans/DeliveryRule.kt [hunk 32]
```diff
@@ -90,6 +90,9 @@ class DeliveryRule : HasOwner {
     )
     var format: DeliveryFormat? = null
 
+    @NotNull
```
Comment (kevinhinterlong) [inline packages/webservice/models/src/main/kotlin/com/yahoo/navi/ws/models/beans/DeliveryRule.kt:93]:
```suggestion
```

File: packages/webservice/app/src/main/resources/db/changelog/changelog.xml [hunk 33]
```diff
@@ -161,6 +161,8 @@
                 ADD isDisabled boolean not null default false;
             ALTER TABLE DeliveryRule 
                 ADD failureCount int not null default 0;
+            ALTER TABLE DeliveryRule    
+                ADD name varchar(255) not null;
         </sql>
     </changeSet>
```
Comment (kevinhinterlong) [inline packages/webservice/app/src/main/resources/db/changelog/changelog.xml:167]: I think we'll have to add a serializer so that the app automatically updates the delivery rule name to its default because it will start out null since we don't have a default in the sql statement.

Otherwise we can have  a default like `"My Delivery Rule"` and keep the not null constraints

```suggestion
        </sql>
    </changeSet>
    <changeSet author="admin" id="5">
        <sql>
            ALTER TABLE DeliveryRule
                ADD name varchar(255);
        </sql>
    </changeSet>
```

**Review by kevinhinterlong — COMMENTED** [review]
Mostly small stuff but there's some issues around the changelog and default name. You can try running
`npm run startElideUI` and `npm run startElideWS` from the app package on master, create a delivery rule, and then switch to this branch and rerun it to see the changelog in action and see that it runs into an error

**Review by kevinhinterlong — APPROVED** [review]

File: packages/reports/addon/components/common-actions/schedule.hbs [hunk 34]
```diff
@@ -2,212 +2,260 @@
...
+	                {{on "click" (fn (mut this.currentDisplayedRule) deliveryRule)}}
+	              >
+                <div class="w-full flex align-items-center">
+                  {{#if deliveryRule.hasDirtyAttributes}}
+                    <div class="schedule__modal-unsaved-rule m-r-10" />
+                  {{/if}}
+                  <div class="schedule__modal-delivery-list-name flex-1">
+                    <p class="is-bold">{{unless deliveryRule.name deliveryRule.defaultName deliveryRule.name}}</p>
+                    {{#if deliveryRule.hasDirtyAttributes}}
+                      <p class="is-italic">&nbsp;unsaved</p>
```
Comment (jkusa) [inline packages/reports/addon/components/common-actions/schedule.hbs:46]:
```suggestion
                      <p class="is-italic">unsaved</p>
```
Do we need the non breaking space?

Non-breaking space
--


<br class="Apple-interchange-newline">

Non-breaking space
--


<br class="Apple-interchange-newline">

**Review by jkusa — APPROVED** [review]
Just the one comment, otherwise 👍
